AttachmentType Registration Enhancements

forgetmenot13579 opened the pull request:
## Register `AttachmentType`s using the builder with consumer pattern.
Today, using the builder pattern creates a `Builder<Object>` by default, making both the `persistent` method and the `initializer` method largely useless, which is especially problematic as this is currently the only way to create an attachment with both persistence and a default value.
![DataType](https://github.com/user-attachments/assets/b6faf91a-bf2d-4d21-a010-c275967f0ae2)
While this can be avoided by explicitly providing type parameters to the static `AttachmentRegistry#builder` method, type parameterizing methods is unintuitive and should be avoided as an API pattern whenever possible.

The solution implemented in this PR is the addition of the `AttachmentRegistry#create(Identifier id, Consumer<Builder<A>>)` method. This allows users to configure the (better type inferred) builder directly.
![Registration](https://github.com/user-attachments/assets/e469c174-3fc1-406e-8ea7-3908b7170641)

forgetmenot13579 committed:
- Add new method AttachmentRegistry#create that allows configuration of the registered attachment type with a builder.
- Migrate existing creation methods to use the new one under the hood for consistency.
- Moves all null checking from AttachmentRegistry to AttachmentRegistryImpl.BuilderImpl (most of them happened there as well already and were thus redundant).

forgetmenot13579 committed:
- Adds the ability to initialize an AttachmentType with only a path, substituting the mod ID of the registrant via an entrypoint.

forgetmenot13579 committed:
- Add registration safety checks

forgetmenot13579 committed:
- Add missing copyright header

modmuss50 commented:
> The motivation for this change is to avoid holding onto the mod ID in magic static strings as much as possible.

This does not seem like the solution to this problem, and barely changes anything as it doesnt solve it in every other place you create an identifier. Just create a static method that returns an identifier so you can do `MyMod.id("path")`. This is a none issue that this change doesnt solve.

forgetmenot13579 committed:
- Remove lazy entrypoint initialization of attachment types.

forgetmenot13579 commented:
I disagree that it is a nonissue, but it is true that solving it here specifically does not make much of a difference when the problem persists across the rest of the ecosystem. I have removed that feature and updated the PR summary accordingly.

forgetmenot13579 committed:
- Import fixes

Syst3ms commented on fabric-data-attachment-api-v1/src/main/java/net/fabricmc/fabric/api/attachment/v1/AttachmentRegistry.java:
```diff
@@ -66,12 +79,9 @@ public static <A> AttachmentType<A> create(Identifier id) {
-		return AttachmentRegistry.<A>builder()
-				.initializer(initializer)
-				.buildAndRegister(id);
+		return create(id, builder -> {
```
nit: use expression lambda

Syst3ms commented on fabric-data-attachment-api-v1/src/main/java/net/fabricmc/fabric/api/attachment/v1/AttachmentRegistry.java:
```diff
@@ -83,10 +93,9 @@ public static <A> AttachmentType<A> createDefaulted(Identifier id, Supplier<A> i
-		Objects.requireNonNull(codec, "codec cannot be null");
-
-		return AttachmentRegistry.<A>builder().persistent(codec).buildAndRegister(id);
+		return create(id, builder -> {
```
likewise, use expression lambda

Syst3ms commented on fabric-data-attachment-api-v1/src/testmod/java/net/fabricmc/fabric/test/attachment/WheelInfo.java:
```diff
@@ -0,0 +1,39 @@
+			Codecs.POSITIVE_FLOAT.fieldOf("tireThickness").forGetter(WheelInfo::tireThickness)
+	).apply(instance, WheelInfo::new));
+
+	public static final AttachmentType<WheelInfo> ATTACHMENT = AttachmentRegistry.create(Identifier.of(AttachmentTestMod.MOD_ID, "wheel_info"),
+			attachment -> attachment.initializer(() -> new WheelInfo(100, 5432, 37))
+					.persistent(WheelInfo.CODEC)
+	);
```
nit: formatting looks weird here

forgetmenot13579 replied:
The important thing is illustrating functionality. I'd be happy to switch the formatting if you prefer any one of these in particular.
```
public static final AttachmentType<WheelInfo> ATTACHMENT = AttachmentRegistry.create(Identifier.of(AttachmentTestMod.MOD_ID, "wheel_info"),
	attachment -> attachment
		.initializer(() -> new WheelInfo(100, 5432, 37))
		.persistent(WheelInfo.CODEC)
);
```
```
public static final AttachmentType<WheelInfo> ATTACHMENT = AttachmentRegistry.create(Identifier.of(AttachmentTestMod.MOD_ID, "wheel_info"),
	attachment -> attachment.initializer(() -> new WheelInfo(100, 5432, 37)).persistent(WheelInfo.CODEC)
);
```
```
public static final AttachmentType<WheelInfo> ATTACHMENT = AttachmentRegistry.create(Identifier.of(AttachmentTestMod.MOD_ID, "wheel_info"),
	attachment -> {
		attachment.initializer(() -> new WheelInfo(100, 5432, 37));
		attachment.persistent(WheelInfo.CODEC);
	}
);
```

Syst3ms replied:
I was thinking either the first and/or putting the arguments on separate lines, as in
```suggestion
	public static final AttachmentType<WheelInfo> ATTACHMENT = AttachmentRegistry.create(
	    Identifier.of(AttachmentTestMod.MOD_ID, "wheel_info"),
	    attachment -> attachment.initializer(() -> new WheelInfo(100, 5432, 37)).persistent(WheelInfo.CODEC)
	);
```
or something like that. Take your pick, this is just a nitpick after all.

forgetmenot13579 replied:
It's in the testmod. I'm just going to leave it as-is.

Syst3ms commented on fabric-data-attachment-api-v1/src/test/java/net/fabricmc/fabric/test/attachment/CommonAttachmentTests.java:
```diff
@@ -33,6 +33,7 @@
 import java.util.function.UnaryOperator;
 
 import com.mojang.serialization.Codec;
+
```
will this build?

Syst3ms commented on fabric-data-attachment-api-v1/src/main/java/net/fabricmc/fabric/api/attachment/v1/AttachmentRegistry.java:
```diff
@@ -43,6 +43,21 @@ public final class AttachmentRegistry {
+	 * @param <A> the type of attached data
+	 * @return the registered {@link AttachmentType} instance
+	 */
+	public static <A> AttachmentType<A> create(Identifier id, Consumer<Builder<A>> consumer) {
```
Since it's intended as a replacement for `builder()`, this pattern should be explicitly documented in the javadoc above.

forgetmenot13579 replied:
Since `builder()` doesn't work as intended anyway, and this *is* an experimental module, would removing or deprecating the existing `builder()` method make sense?

Syst3ms replied:
Removing is definitely too much. Honestly, I knew of the issue already, I just thought passing the type parameters was alright. There's absolutely no need to make a breaking change, considering the existing method works, even if it's a bit clunky.

Deprecation sounds a bit more reasonable, but I think @modmuss50 knows the policy better on that front.

modmuss50 replied:
Deprecating the old methods (without forRemoval) seems fine, there is no benefit to removing them.

Syst3ms reviewed (changes requested):
Mostly nitpicks, this is a reasonable change. I think the entrypoint thing was well out of scope, though. This won't be too big of a hassle to merge into attachment sync.

forgetmenot13579 committed:
- Update javadoc.

forgetmenot13579 committed:
- Use expression lambdas.

forgetmenot13579 committed:
- Fix style checks

forgetmenot13579 committed:
- More style fixes

forgetmenot13579 committed:
- Fix line endings

modmuss50 commented on fabric-data-attachment-api-v1/src/testmod/java/net/fabricmc/fabric/test/attachment/WheelInfo.java:
```diff
@@ -0,0 +1,40 @@
+import net.fabricmc.fabric.api.attachment.v1.AttachmentRegistry;
+import net.fabricmc.fabric.api.attachment.v1.AttachmentType;
+
+public record WheelInfo(float wheelDiameter, float tireDiameter, float tireThickness) {
```
This class is never loaded, so we never actually run this test code. I would suggest replacing this with a unit test in CommonAttachmentTests if possible.

forgetmenot13579 replied:
This was mostly included to show the API usage, rather than as an actual test, but I have moved it to the test package so it does get registered. I've also updated the existing tests to use the new API.

forgetmenot13579 committed:
- Move WheelInfo and initialization to test package
- Deprecate AttachmentRegistry#builder
- Update existing tests to use #create rather than #builder
